### DISCUSSION TITLE
Using Torchdistill with MMDetection

### DISCUSSION THREAD

**Discussion by ai21mtech13002** [yoshitomo-matsubara/torchdistill#264] [Q&A] [answered]
Hi Yoshitomo,

Can I use models trained in MMDetection with TorchDistill to run experiments?

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @ai21mtech13002 

Yes, you can.
If the trained model weights are a form of `state_dict`, you can specify the file path as `ckpt`

P.S. Could you also close [the previous discussion](https://github.com/yoshitomo-matsubara/torchdistill/discussions/256)?

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

Actually MMDetection uses config files to define and train models. So I am a bit confused how the same can be integrated with TorchDistill.

**Reply by yoshitomo-matsubara** [reply]
You will write a function like
```python

@register_model_func
def your_mmdetection_model(config_file_path):
    config = ... # read the config file
    model = the_mmdetection_function_or_class(config_file)
    return model
```

and then the model can be defined in the yaml like 
```yaml
model:
  name: 'your_mmdetection_model'
  params:
    config_file_path: ...
```

**Reply by ai21mtech13002** [reply]
Thanks, I will try this.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

The mmdetection model is working on the DOTA dataset. The model takes the annotations in the DOTA format and not the COCO format. How do I make this work with torchdistill?

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

Technically, you can use datasets in non-COCO formats with torchdistill, but then you will need to modify the exaple script, [`examples/object_detection.py`](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/examples/object_detection.py) in this case (especially [evaluate function](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/examples/object_detection.py#L101-L155) should be replaced with a DOTA-format-compatible evaluate function)

**Reply by ai21mtech13002** [reply]
What other modifications would be required in the config file? I am enclosing a sample annotation file for reference
[P0000.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10578193/P0000.txt)

P.S. The annotations are for Rotated Bounding Boxes. I also considered converting them to COCO annotations , but COCO does not support Rotated Bounding Box annotations.

**Reply by ai21mtech13002** [reply]
Also, mmdetection carries out the following preprocessing prior to passing  the batch through the model.
<img width="603" alt="Screenshot 2023-02-03 at 4 32 21 PM" src="https://user-images.githubusercontent.com/104313548/216587771-432e7e5e-102e-4ddc-a2bf-d2810850e90a.png">

These are defined as classes in the mmdetection repository.
<img width="617" alt="Screenshot 2023-02-03 at 5 37 02 PM" src="https://user-images.githubusercontent.com/104313548/216600175-328b1d6a-88f8-48d7-8284-b9c3b1962fdd.png">



How do I account for the same in Torchdistill?

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

You will register and fine datasets and model configs like you did [here](https://github.com/yoshitomo-matsubara/torchdistill/discussions/254), potentially + your custom [forward_proc](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/coco2017/multi_stage/ft/custom_fasterrcnn_resnet18_fpn_from_fasterrcnn_resnet50_fpn.yaml#L149) to re-select what you want to feed to the model.


> P.S. The annotations are for Rotated Bounding Boxes. I also considered converting them to COCO annotations , but COCO does not support Rotated Bounding Box annotations.

I am not very familiar with Rotated Bounding Box annotations, MMDetection or their pipeline designs either. If the training pipeline just performs dataset pre-processing, you can modify the dataset model class (and introduce your custom `collate` and/or `batcher` if necessary) to work with torchdistill.

**Reply by ai21mtech13002** [reply]
Thanks Yoshitomo. I will look into this.
